bricks/stm32/main.c

```diff
@@ -95,6 +96,72 @@ static const pbsys_user_program_callbacks_t user_program_callbacks = {
     .stdin_event = user_program_stdin_event_func,
 };
 
+typedef struct _pb_reader_user_data_t {
+    const byte *beg;
+    const byte *cur;
+    const byte *end;
+} pb_reader_user_data_t;
+
+mp_uint_t pb_reader_user_data_readbyte(void *data) {
+    pb_reader_user_data_t *blob = (pb_reader_user_data_t *)data;
+    if (blob->cur < blob->end) {
+        return *blob->cur++;
+    } else {
+        return MP_READER_EOF;
+    }
+}
+
+static void pb_reader_user_data_close(void *data) {
+    (void)data;
+}
+
+const uint8_t *pb_reader_user_data_readchunk(void *data, size_t len) {
+    pb_reader_user_data_t *blob = (pb_reader_user_data_t *)data;
+    const uint8_t *ptr = blob->cur;
+    blob->cur += len;
+    return ptr;
+}
+
+static void pb_reader_user_data_new(mp_reader_t *reader, pb_reader_user_data_t *data, const byte *buf, size_t len) {
+    data->beg = buf;
+    data->cur = buf;
+    data->end = buf + len;
+    reader->data = data;
+    reader->readbyte = pb_reader_user_data_readbyte;
+    reader->readchunk = pb_reader_user_data_readchunk;
+    reader->close = pb_reader_user_data_close;
+}
+
+static void do_execute_raw_code(mp_module_context_t *context, const mp_raw_code_t *rc, const mp_module_context_t *mc) {
```

Comment:
We should mention that this is copied from MicroPython in case it is made public in the future and there is a possibility of sharing the code.